### PR TITLE
perf: strftime partitioner caching

### PR DESCRIPTION
**Pull request by domodwyer**
This PR is based on two observations:

* Writes tend to be batched, and within those batches, timestamps repeat (often multiple rows have the same timestamp) - caching the generated string for these values saves us from generating the same string repeatedly.

* Writes using the default partitioning scheme of `YYYY-MM-DD` only utilise the higher-order components of a timestamp - the values of the hours, minutes, seconds, etc have no bearing on the output partition key - normalising the timestamps by discarding the lower-order values increases the cache hit rate.

When normalising the timestamps for the default partition template, the cache hit rate / re-use of prior generated strings **hits 100%** for writes that span less than 6 days, regardless of the number of rows. We generate one string per day, and keep reusing that after.

This PR leverages those two points above to increase the partitioner performance by ~500% for the default YYYY-MM-DD partitioning strategy, and ~400% for custom strftime formatting templates. All sampled workloads and write batch sizes are faster with this PR.

---

* perf: strftime partitioner caching (0b5b6a8e1)
      
      This commit extracts the strftime formatting logic into it's own type,
      and implements a small ring-buffer based cache containing the last 5
      observed timestamps (lazily initialised).
      
      This optimisation leverages the fact that the typical write to IOx is a
      batch containing many hundreds or thousands of rows - often these rows
      are measurements of multiple variables at the same timestamp; for
      example, a metric scrape system will periodically read a set of metrics
      and assign them all the same timestamp (the "scrape timestamp").
      
      Because of the above, batches often contain multiple rows with the same
      timestamp, so we can reduce the overhead by cacheing the resulting
      partition key value for any given timestamp, eliminating the need to
      re-compute it for these successive identical values. We retain the last
      5 observed timestamps (FIFO) to provide a degree of "look back".
      
      Alone the above is effective for measurements all with the same exact
      timestamp, often a subset of a batch. However a further optimisation is
      possible: because the default partitioning scheme (YYYY-MM-DD) operates
      at a granularity of days, the timestamp precision can be reduced
      (discarding hours, minutes, seconds, etc) without effecting the
      resulting partition key. Therefore when the default partitioning scheme
      is used, this commit will normalise timestamps to match this reduced
      precision before caching, causing the cache hit & string re-use rate to
      rise to 100% for batches containing measurements that span < 6 days.
      
      This brings a ~5x improvement against a modest batch size of 1,000
      lines, showing improvement across all batch sizes and partitioning
      schemes (default & custom). I'd expect the performance improvement to be
      even greater for larger batches.

### CODE REVIEW

File: mutable_batch/src/payload/partition/strftime.rs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+/// the value, without changing the formatter output (it still produces the same
+/// string). This in turn causes any timestamp from the same day to be a cache
+/// hit with any prior value for the same day, regardless of "time" portion of
+/// the timestamp.
+///
+/// Combined with the above cache, this raises the cache hit rate to ~100% for
+/// write batches that span less than 6 days, effectively amortising the cost of
+/// timestamp formatting to O(1) for these very common batches.
+#[derive(Debug)]
+pub(super) struct StrftimeFormatter<'a> {
```
Comment (domodwyer) [inline mutable_batch/src/payload/partition/strftime.rs:127]: This type is responsible for the formatting of strftime portions of templates, and contains an embedded cache of up to 5 most recently observed timestamps.

File: mutable_batch/src/payload/partition/strftime.rs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    fn maybe_reduce(&self, timestamp: i64) -> i64 {
+        if !self.is_ymd_format {
+            return timestamp;
+        }
+        // Don't map timestamps less than the value we would subtract.
+        if timestamp < DAY_NANOSECONDS {
+            return timestamp;
+        }
+        timestamp - (timestamp % DAY_NANOSECONDS)
+    }
```
Comment (domodwyer) [inline mutable_batch/src/payload/partition/strftime.rs:233]: Here we apply a precision reduction optimisation described in the PR description, to increase the cache hit rate using the default YYYY-MM-DD template.

File: mutable_batch/src/payload/partition/strftime.rs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    {
+        // Optionally apply the default format reduction optimisation.
+        let timestamp = self.maybe_reduce(timestamp);
+
+        // Check if this timestamp has already been rendered.
+        if let Some(v) = self.values.find(|(t, _v)| *t == timestamp) {
+            // It has! Re-use the existing formatted string.
+            out.write_str(&v.1)?;
+            return Ok(());
+        }
```
Comment (domodwyer) [inline mutable_batch/src/payload/partition/strftime.rs:187]: We re-use previously generated strings if they've been cached.

File: mutable_batch/src/payload/partition/strftime.rs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+        // Obtain a mutable reference to the next item to be replaced, re-using
+        // the string buffer within it to avoid allocating (or initialising it
+        // if it was not yet initialised).
+        let buf = self.values.next_slot();
+
+        // Reset the slot value
+        buf.0 = timestamp;
+        buf.1.clear();
```
Comment (domodwyer) [inline mutable_batch/src/payload/partition/strftime.rs:196]: And re-use previously allocated strings if we're evicting a key from the cache.

File: mutable_batch/src/payload/partition/strftime.rs [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        /// formatter, therefore this test asserts the following property:
+        ///
+        ///     For any timestamp and formatter, the output of this type must
+        ///     match the output of chrono's formatter, after URL encoding.
+        ///
+        /// Validating this asserts correctness of the wrapper itself, assuming
+        /// chrono's formatter produces correct output. Note the encoding is
+        /// tested in the actual partitioner module.
+        #[test]
+        fn prop_differential_validation(
```
Comment (domodwyer) [inline mutable_batch/src/payload/partition/strftime.rs:324]: This test ensures the cache & reduction optimisations never effect the output.

File: mutable_batch/src/payload/partition/strftime.rs [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    use super::*;
+
+    #[test]
+    fn test_default_formatter() {
+        let template = TablePartitionTemplateOverride::default();
+        let expect = template.parts().collect::<Vec<_>>();
+
+        // If this assert fails (and it probably shouldn't!) then you may want
+        // to consider changing the special case optimisation above.
+        assert_matches!(expect.as_slice(), &[TemplatePart::TimeFormat(YMD_SPEC)]);
```
Comment (savage-engineer) [inline mutable_batch/src/payload/partition/strftime.rs:251]: A good old assert with complimentary helpful suggestions 🥇

**Review by savage-engineer — APPROVED** [review]

**Pull request merged**